### PR TITLE
Support LPC55S36 in Zephyr

### PR DESCRIPTION
**Pull request by NeilChen93**

### CODE REVIEW

**Comment by NeilChen93** [comment]
Hi all, this PR is prepared for LPC55S36, please help review, thanks a lot. And the file pinctrl_soc.h in location zephyr\soc\arm\nxp_lpc\lpc55xxx\pinctrl_soc.h need to be updated, because LPC55S36’s analog switch input control bit name is different with other LPC55xx chips. Daniel will be responsible to update it. Then LPC55S36's hello_world example can build and run successfully.

**Comment by dleach02** [comment]
@NeilChen93, we need to clean up the commit history here. The "merge" of main commit needs to go away where your branch would use a rebase from main to get your branch up to date.

**Review by dleach02 — CHANGES_REQUESTED** [review]
Use rebase to update branch with main.

File: soc/arm/nxp_lpc/lpc55xxx/soc.c [hunk 1]
```diff
@@ -59,7 +59,8 @@ const pll_setup_t pll0Setup = {
 static ALWAYS_INLINE void clock_init(void)
 {
 #if defined(CONFIG_SOC_LPC55S06) || defined(CONFIG_SOC_LPC55S16) || \
-	defined(CONFIG_SOC_LPC55S28) || defined(CONFIG_SOC_LPC55S69_CPU0)
+	defined(CONFIG_SOC_LPC55S28) || defined(CONFIG_SOC_LPC55S36) || \
+	defined(CONFIG_SOC_LPC55S69_CPU0)
     /*!< Set up the clock sources */
```
Comment (dleach02) [inline soc/arm/nxp_lpc/lpc55xxx/soc.c:64]: Coding standard for Zephyr uses tabs instead of spaces

**Comment by mmahadevan108** [comment]
Please add the pinctrl related updates made by @danieldegrasse.

File: boards/arm/lpcxpresso55s36/Kconfig.defconfig [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# LPCXpresso55S36 board
+
+# Copyright (c) 2022 NXP
```
Comment (mmahadevan108) [inline boards/arm/lpcxpresso55s36/Kconfig.defconfig:3]: NXP guidance on Copyright is to drop the `(c)`

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36_common.dtsi [hunk 3]
```diff
@@ -0,0 +1,114 @@
+/*
```
Comment (mmahadevan108) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36_common.dtsi:1]: Do we need a `lpcxpresso55s36_common.dtsi` file. Seems like this can be move to `lpcxpresso55s36.dtsi`.

**Comment by danieldegrasse** [comment]
@NeilChen93 Do you want me to open a separate PR with the updates for the pin control driver, or integrate those changes here?

File: boards/arm/lpcxpresso55s36/Kconfig.defconfig (hunk 2 above)
Comment (NeilChen93) [reply boards/arm/lpcxpresso55s36/Kconfig.defconfig:3]: Thanks for reminder.

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36_common.dtsi (hunk 3 above)
Comment (NeilChen93) [reply boards/arm/lpcxpresso55s36/lpcxpresso55s36_common.dtsi:1]: I will try to move this file into lpcxpresso55s36.dtsi

**Comment by NeilChen93** [comment]
Hi @danieldegrasse, it is both OK for me. I prefer you to open a separate PR.

**Comment by danieldegrasse** [comment]
> Hi @danieldegrasse, it is both OK for me. I prefer you to open a separate PR.
@NeilChen93 #48193 should contain the required updates for the LPC55s3x family pin control driver

**Comment by NeilChen93** [comment]
Hi all, I cherry pick daniel's commit into this PR. And LPC55S36 examples should can build and run now, please help review and check again. Thanks a lot.

**Review by dleach02 — CHANGES_REQUESTED** [review]
@NeilChen93, the image for the board is a LPC55S16 board and not the LPC5536 board

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc [hunk 4]
```diff
@@ -29,6 +30,7 @@ config SOC_LPC55S16
 	select CLOCK_CONTROL
 	select HAS_MCUX_IAP if !TRUSTED_EXECUTION_NONSECURE
 	select HAS_MCUX_MCAN
+	select HAS_MCUX_RNG
```
Comment (dleach02) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:33]: We should focus this PR only on the LPC55S36. If we want to add the select of RNG to other platforms lets do that in a standalone PR that also connects the entropy driver.

File: soc/arm/nxp_lpc/lpc55xxx/soc.c (hunk 1 above)
Comment (dleach02) [reply soc/arm/nxp_lpc/lpc55xxx/soc.c:64]: This applies to the comment block as well

**Comment by dleach02** [comment]
@NeilChen93, please address the CI failures

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+};
+
+&peripheral {
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	syscon: syscon@0 {
+		compatible = "nxp,lpc-syscon";
+		reg = <0x0 0x1000>;
+		label = "SYSCON";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:73]: remove `label`

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+		#clock-cells = <1>;
+	};
+
+	iap: flash-controller@34000 {
+		compatible = "nxp,lpc-iap";
+		label = "FLASH_IAP";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:79]: remove `label` property

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+		reg = <0x34000 0x1000>;
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		flash0: flash@0 {
+			compatible = "soc-nv-flash";
+			label = "MCUX_FLASH";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:86]: remove `label` property

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+			compatible = "nxp,lpc-iocon-pio";
+			reg = <0x80 0x80>;
+		};
+	};
+
+	gpio0: gpio@0 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x8c000 0x278c>;
+		interrupts = <4 2>,<5 2>,<6 2>,<7 2>;
+		label = "GPIO_0";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:132]: remove `label` property

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <0>;
+	};
+
+	gpio1: gpio@1 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x8c000 0x278c>;
+		interrupts = <32 2>,<33 2>,<34 2>,<35 2>;
+		label = "GPIO_1";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:142]: remove `label` property

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+		gpio-controller;
+		#gpio-cells = <2>;
+		port = <1>;
+	};
+
+	gpio2: gpio@2 {
+		compatible = "nxp,lpc-gpio";
+		reg = <0x8c000 0x278c>;
+		#interrupts = <32 2>,<33 2>,<34 2>,<35 2>;
+		label = "GPIO_2";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:152]: remove `label` property

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+		#gpio-cells = <2>;
+		port = <2>;
+	};
+
+	flexcomm0: flexcomm@86000 {
+		compatible = "nxp,lpc-flexcomm";
+		reg = <0x86000 0x1000>;
+		interrupts = <14 0>;
+		clocks = <&syscon MCUX_FLEXCOMM0_CLK>;
+		label = "FLEXCOMM_0";
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:163]: remove `label` property

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 12]
```diff
@@ -0,0 +1,245 @@
+/*
```
Comment (galak) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:1]: remove `label` property from all nodes in here

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36 [hunk 13]
```diff
@@ -0,0 +1,27 @@
+# NXP LPC55S36 platform configuration options
+
+# Copyright 2022 NXP
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_LPC55S36
+
+config SOC
+	default "lpc55S36"
+
```
Comment (galak) [inline soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36:10]: remove PINMUX_MCUX_LPC, GPIO_MCUX_LPC, and UART_MCUX_FLEXCOMM.  They aren't needed.

**Review by galak — CHANGES_REQUESTED** [review]
* Remove `label` usage from soc dtsi
* Some Kconfig updates

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi [hunk 14]
```diff
@@ -0,0 +1,51 @@
+/*
+ * NOTE: File generated by lpc_cfg_utils.py
+ * from LPC55S36.mex
+ *
+ * Copyright 2022, NXP
```
Comment (yvanderv) [inline boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:5]: Why a comma here?

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+		compatible = "mmio-sram";
+		reg = <0x4000000 DT_SIZE_K(16)>;
+	};
+
+	sram0: memory@20000000 {
+		compatible = "mmio-sram";
+		reg = <0x20000000 DT_SIZE_K(32)>;
+	};
+
+	sram2: memory@20008000 {
```
Comment (yvanderv) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:47]: Why not SRAM1?

Comment (dleach02) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi:47]: @NeilChen93, lets do something similar that was done for the other LPC parts (with related changes to the ns dtsi):

	/* lpc55_3x Memory configurations:
	 *
	 * LPC55x36: RAMX: 16K, SRAM0: 16K, SRAM1: 16K, SRAM2: 32K, SRAM3: 32K, SRAM4: 16K
	 */
	sramx: memory@4000000 {
		compatible = "mmio-sram";
		reg = <0x04000000 DT_SIZE_K(16)>;
	};
	sram0: memory@20000000 {
		compatible = "mmio-sram";
		reg = <0x20000000 DT_SIZE_K(16)>;
	};
	sram1: memory@20004000 {
		compatible = "mmio-sram";
		reg = <0x20004000 DT_SIZE_K(16)>;
	};
	sram2: memory@20008000 {
		compatible = "mmio-sram";
		reg = <0x20008000 DT_SIZE_K(32)>;
	};
	sram3: memory@20010000 {
		compatible = "mmio-sram";
		reg = <0x20010000 DT_SIZE_K(32)>;
	};
	sram4: memory@20018000 {
		compatible = "mmio-sram";
		reg = <0x20018000 DT_SIZE_K(16)>;
	};

File: dts/arm/nxp/nxp_lpc55S36_ns.dtsi [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+			ranges = <0x4000000 0x4000000 0x20000000>;
+		};
+
+		peripheral: peripheral@40000000 {
+			ranges = <0x0 0x40000000 0x10000000>;
+		};
+	};
+};
+
+#include "nxp_lpc55S3x_common.dtsi"
```
Comment (dleach02) [inline dts/arm/nxp/nxp_lpc55S36_ns.dtsi:19]: After this line, add the following:

/*
 * lpc55S_36:
 * Combine SRAM0, SRAM1, SRAM2, SRAM3, SRAM4 for total of 112K RAM
 */
&sram0 {
	compatible = "mmio-sram";
	reg = <0x20000000 DT_SIZE_K(112)>;
};

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc (hunk 4 above)
Comment (NeilChen93) [reply soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:33]: Hi @dleach02 , LPC55S36 don't support RNG, so I removed LPC55XXX common config HAS_MCUX_RNG in file "soc/arm/nxp_lpc/lpc55xxx/Kconfig.series", and add the select of RNG separately for other platforms. Is there any other better way to remove RNG support for LPC55S36? Thanks a lot.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 5 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 6 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 7 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 8 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 9 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 10 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 11 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi]: Already removed it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 12 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi:1]: Already removed it.

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36 (hunk 13 above)
Comment (NeilChen93) [reply soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36:10]: Already removed it.

File: boards/arm/lpcxpresso55s36/doc/index.rst [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+- `LPC55S36 Datasheet`_
+- `LPC55S36 User Manual`_
+- `LPCXpresso55S36 Website`_
+- `LPCXpresso55S36 User Manual`_
+- `LPCXpresso55S36 Development Board Design Files`_
+
+Supported Features
+==================
+
+The lpcxpresso55s36 board configuration supports the following
```
[the remark this answers is not in the document]
Comment (NeilChen93) [reply boards/arm/lpcxpresso55s36/doc/index.rst]: Updated it.

File: boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi (hunk 14 above)
Comment (NeilChen93) [reply boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi]: Updated it.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 15 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi:53]: Thanks dleach.

File: dts/arm/nxp/nxp_lpc55S36_ns.dtsi (hunk 16 above)
Comment (NeilChen93) [reply dts/arm/nxp/nxp_lpc55S36_ns.dtsi:19]: Got it, thanks for reminder.

File: dts/arm/nxp/nxp_lpc55S36_ns.dtsi [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+
+/*
+ * lpc55S_36:
+ * Combine SRAM0, SRAM1, SRAM2, SRAM3, SRAM4 for total of 112K RAM
+ */
+&sram0 {
+	compatible = "mmio-sram";
+	reg = <0x20000000 DT_SIZE_K(112)>;
+};
```
Comment (dleach02) [inline dts/arm/nxp/nxp_lpc55S36_ns.dtsi:28]: need newline after

File: soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc (hunk 4 above)
Comment (dleach02) [reply soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:33]: okay. That makes sense.

File: dts/arm/nxp/nxp_lpc55S36_ns.dtsi (hunk 18 above)
Comment (dleach02) [reply dts/arm/nxp/nxp_lpc55S36_ns.dtsi:28]: @NeilChen93, you need a newline at the end of the file. This will clean up one of the conformance test errors in CI

**Comment by NeilChen93** [comment]
Hi @dleach02 , @galak , I already updated according to your comments and this PR has passed all CI test, so please help check again and approve.

**Review by dleach02 — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Review by mmahadevan108 — APPROVED** [review]

**Review by danieldegrasse — APPROVED** [review]

**Review by yvanderv — COMMENTED** [review]
Looks good to me
